README.md

```diff
@@ -422,6 +432,30 @@ When storing logs on disk, we recommend using Logrotate in truncate mode.
 
 You can use [jq](https://stedolan.github.io/jq/) to traverse JSON log files, when you're not running Kiev in *development mode*.
 
+### Pre-tagging `tree_path`
+
+Sometimes automtic management of `tree_path` via built-in middlewares and extensions doesn't cut it.
+
+For example, when sending a message to an MQ topic (e. g. in RabbitMQ or SNS) there no way to assign distinct `tree_path`s to individual subscribers. And since all subscribers receive identical messages, they log identical `tree_path`s, and if they do loggable actions inside (perform requests or add jobs) you have no way of knowing which of all subscribers did this. This makes `tree_path` much less useful in these scenarios. But it's fixable.
+
+To keep the request trace clear and unambiguous, you'll have to choose unique identifiers for all susbcriber groups you may want to identify (e. g. consumption types, if each message is consumed at most once in every way). Then add a suffix for `tree_path` using a middleware inside the subscriber code.
+
+* Keep the suffix short, preferably a single character.
```

Comment:
there is specific encoding behind tree path. If we can not repeat exact functionality let's skip it. Otherwise it will create more confusion than value.

Base52 flavor is behind encoding. We use only alphabet to make sure it is lexicographically sortable. For sync requests we use even numbers e.g. 0,2,4, which translates to A,C,E. For async numbers we use odd numbers. By looking at tree path I can tell exact path of request. 

For example, `ABC` decodes too:

```
A--(schedule async job)--B--(?)-->
                         |
                         |--(synchronous request)-->C
```

`tree_path` is "poorman's" distributed tracing scheme.

Reply:
On the contrary, having entries from multiple sources with identical `tree_path`s is misleading and is even more confusing than having no `tree_path` at all. If it exists, it has to be usable. Sure, we can strip it entirely if it's that unreliable, but we **can** make it work.

Conventions below suggest a good way to pick a suffix and maintain sortability in commonly used encodings (though I don't think anyone will use this with something other than UTF-8): `~` lies farther than `z` and is lexicographically "larger" than any entries in the same branch logged with leters-only path.

So, for instance by having pre-tagging with prefixes such as `~1` and `~2` on two different consumers we get:

* all message consumers' logs conveniently grouped in a trace at the bottom of the sender request's branch
* subrequests identifiable by-subscriber (assuming prefixes are unique among different handlers of the same message, with can be done by choosing them carefully)

Reply:
Be aware there is already confusion in who is responsible for what, and this thing will increase it more.

- request_depth increased on receiver
- tree_path increased on sender (as of now)

and after this it will be

- request_depth increased on receiver
- tree_path increased on sender by default
- tree_path increased on receiver in case of Shoryuken

Also I'm not quite follow all the parts of the system. You are saying one message can be processed by more than one subscriber?

Reply:
> You are saying one message can be processed by more than one subscriber?

Precisely. Hence, `tree_path` cannot be increased on sender in this case.

But this isn't typical usage of Shoryuken, typical usage is very Sidekiq-like: send one payload, have it handled once, done. So I did not require a tag in `Kiev::Shoryuken.enable`.

This is only required for unusual setups, such as the one I'm handling at the moment, with multiple SQS queues in multiple services on the same logging infrastructure subscribed to a common SNS topic.

Reply:
What if those subscribers are on different machines. How they will coordinate incrementing the flag?

Reply:
They **are** on different machines. And yeah, I don't see a way to make it work the usual way, so I made this workaround, so different subscribers "increment the flag by their own amounts", so to speak.

Reply:
But how this will help?

Message with `AB` tree_path comes to **1** machine. It produces `AB~1`.
Message with `AB` tree_path comes to **2** machine. It produces `AB~1`.

Or if each machine will produce the same addon every time, like

Message with `AB` tree_path comes to **1** machine. It produces `AB~1`.
Message with `AB` tree_path comes to **2** machine. It produces `AB~2`.

They are already out of scheme, they are sortable but order doesn't necessary the same as it was scheduled.

If they are on different machines they will help different `application` value - this is how you can differentiate logs with the same path on different machines.

What I try to say - you are increasing complexity of the system, without actually preserving "value" of tree_path. Abaddon tree_path or come up with more elegant scheme.

Reply:
> Or if each machine will produce the same addon every time

This. For common codebases it could be an infrastructure thing. I'm dealing with different subscriber codebases, so I can assign one tag for each (in initializers or service-scoped configuration files).

> They are already out of scheme, they are sortable but order doesn't necessary the same as it was scheduled.

Producer only sends one message, so there is no scheduling order to speak of. On this particular level, sorting serves more for grouping all entries about the message and subgrouping them by subscriber.

> If they are on different machines they will help different `application` value - this is how you can differentiate logs with the same path on different machines.

Except if they each make another subrequest, at which point it becomes impossible to tell which consumer did this.

Reply:
> What I try to say - you are increasing complexity of the system, without actually preserving "value" of tree_path. Abaddon tree_path or come up with more elegant scheme.

The ideal solution would be if SNS could somehow number each copy of the message. But it can't do that. So nothing can be done on sender's side, as fanout happens in a place that we have no control over.
So subscribers either have to use:
   * a common subrequest counter to use for `tree_path` (and maintaining that system-wide piece of state introduces a whole host of problems, so it's out of question)
   * ...or each use a predefined value for said counter, so each appends the same letter to `tree_path` for every message it receives, which is...

...roughly the solution implemented in that middleware. Except the "predefined value" is free to be chosen by developers of any particular system.

So I disagree. Of all considered solutions this one has the least impact on complexity and the system overall (no infrastructure changes, minimal code changes, producer side is not impacted). And it actually does make the trace unambiguous, which is the whole idea behind `tree_path`: given a set of events that happened, indicate what events caused what, both directly and indirectly.

Reply:
Hm... actually common prefix `~` for all of consumers' suffixes might not be necessary, since the SNS message is sent with data from Kiev's `SubrequestHelper.payload` and is assigned a unique request-wide `tree_path`. So `~` will have no effect anyway, as the last letter of `tre_path` in the messages can fill its role.
Uniqueness, however, is still necessary.

Reply:
So `Kiev::Shoryuken.tag_tree_path(self, "B")` (for machine 1) and `Kiev::Shoryuken.tag_tree_path(self, "D")` (for machine 2) would be enough?

Can we remove text about 
> By convention, use `~` to make suffixed entries appear at the bottom of its parent branch or `-` for them to appear at the top.
>  * You can also choose other symbols by looking up character codes before `A` and after `z` (note the case).
> * If your subscribers can be sensibly ordered, having corresponding lexicographic ordering of their suffixes might help.

It more confuse than help. 

Also can we come up with better name for `tag_tree_path`. Calling for naming wizard @lipanski

Reply:
I've been rewriting this part entirely, will be here in a few minutes :)